### PR TITLE
feat(gatsby): switch to different caching mechanism backed by fs

### PR DESCRIPTION
**Pull request by DSchau**
This is a _very_ WIP implementation of a different caching mechanism.

The general idea is that we switch to files, which are the backing mechanism for the cache. Additionally, we pass a unique (based on plugin name) cache instance to each plugin, rather than re-using the same cache file for every plugin.

I think this is an improvement, but still need to test this quite a bit more. It doesn't fail on gatsby-bench, and performs similar to what we already had.

Next thing I want to do is check out [nedb](https://github.com/louischatriot/nedb) which might be the best of both worlds.

### CODE REVIEW

File: packages/gatsby/package.json [hunk 1]
```diff
@@ -32,6 +32,8 @@
     "babel-plugin-remove-graphql-queries": "^2.0.2-rc.3",
     "better-queue": "^3.8.6",
     "bluebird": "^3.5.0",
+    "cache-manager": "^2.9.0",
+    "cache-manager-fs": "^1.0.6",
```
Comment (DSchau) [inline packages/gatsby/package.json]: There's also a [redis cache manager](https://www.npmjs.com/package/cache-manager-redis) 👍 

I haven't thought through how to swap these out yet (gatsby-plugin-cache?) but that's something we'll want to think about.

**Comment by DSchau** [comment]
The cache ends up looking like this:

screen shot 2018-09-21 at 3 56 01 pm

**Comment by DSchau** [comment]
Note: this seemed to fix the issue on gatsby-bench, but subsequent re-builds tended to be slower than they are now, so more to do here.

File: packages/gatsby/src/utils/cache.js [hunk 2]
```diff
@@ -1,73 +1,44 @@
...
+const path = require(`path`)
+
+const MAX_CACHE_SIZE = 250
+
+class Cache {
+  constructor(folderName = `db`, store = fsStore) {
+    this.folderName = folderName
+
+    const caches = [
+      { store: `memory`, max: MAX_CACHE_SIZE },
```
Comment (DSchau) [inline packages/gatsby/src/utils/cache.js]: Ideally, there'd probably be some way to opt into this at the plugin level.

**Comment by eLod** [comment]
just checked, i can confirm the pr fixes my bench and also it does pass the query generation for 3x content, though it OOMs after bootstrap (which is most likely because of the redux state `JSON.stringify`, so a different issue)

File: packages/gatsby/src/utils/cache.js [hunk 3]
```diff
@@ -1,73 +1,47 @@
...
+  init() {
+    fs.ensureDirSync(this.directory)
+
+    const caches = [
+      { store: `memory`, max: MAX_CACHE_SIZE },
+      {
+        store: this.store,
+        options: {
+          path: this.directory,
+          ttl: ONE_HOUR,
```
Comment (DSchau) [inline packages/gatsby/src/utils/cache.js]: I am not sure if this is too high or too low 🤷‍♂️

File: benchmarks/markdown/page-template.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+const faker = require(`faker`)
+const matter = require(`gray-matter`)
+
+const NUM_ROWS = process.env.NUM_ROWS || 25;
```
Comment (DSchau) [inline benchmarks/markdown/page-template.js]: Note: swapping this to even 100 or so fails the build with OOM.

e.g.

``bash
NUM_ROWS=100 yarn build
```

will reliably fail, so this is a good, general-purpose test case/template.

**Comment by DSchau** [comment]
Whoa, so I'm pretty surprised with the performance here. This _tends_ to be slower in the bootstrap phase, and less queries are run (averaging just under 200/s with this implementation whereas it's more lower/mid 200s with existing)--I believe because the caching is a little bit slower with multiple implementations, i.e. cache.get/cache.set will be slower. However, it seems like it speeds up other areas of the Gatsby build process.

## Steps

1. Open a terminal tab `cd`'ed into Gatsby
1. Run `yarn watch --scope gatsby`
1. Open a new tab, and `cd benchmarks/markdown`
1. `gatsby-dev --set-path-to-repo path-to-gatsby-repo`
1. `gatsby-dev --scan-once --packages gatsby`
1. `rm -rf .cache` # only do this step if cold cache
1. `time yarn build`
  - Use the raw numbers, not the numbers reported from Gatsby; figured that was more fair

to compare without the cache, run `rm -rf node_modules && yarn` and follow the rest of the steps after `gatsby-dev`. I would appreciate at least one other set of 👀 to make sure everything makes sense and can be replicated.

## Performance Metrics

I used the [markdown benchmark](./benchmarks/markdown) to test, and used gatsby-dev-cli to test the new vs. old caching mecanism.

A "cold" cache is enabled by `rm -rf .cache` before running `time yarn build`

A "warm" cache is persisting the `.cache` folder between runs

|Description|Test|Test|Test|Test|Test|Average|
|:---------:|----|----|----|----|----|-------|
|cache-manager (cold)|47.96s|54.26s| 59.30s| 59.04s| 58.05s|55.72s|
|cache-manager (warm)|19.53s|18.60s|21.94s|21.72s|19.85s|20.33s|
|existing (cold)|65.84s|57.26s|68.19s|63.05s|56.91s|62.25s|
|existing (warm)|24.39s|25.11s|23.85s|24.30s|24.56s|24.42s|

I'm seeing `10%` improvement in cold cache and ~`16.7%` improvement in warm cache behavior. Not nothing, and it seems to fail more slowly with OOM which is obviously a big win.

**Comment by DSchau** [comment]
So also to demonstrate the benefit of this PR relating to OOM, run the following.

## Steps

```bash
NUM_ROWS=50 yarn build
```

this will add more Markdown content to the files, therefore more quickly OOM'ing.

Swapping to the new cache (see above with gatsby-dev-cli), you can run

```bash
NUM_ROWS=50 yarn build
```

and it should succeed. It does seem to fail around 100, however, so there's still more to do here.

**Comment by KyleAMathews** [comment]
This is looking fantastic!

One test failing btw

**Comment by DSchau** [comment]
@KyleAMathews I think some of the test runs were _too_ quick, but with some help from @pieh I think we're getting closer. Want to test it on his machine tomorrow (which OOM'ed intermittently with the new cache strategy), but I think this is starting to look better, albeit not quite as quick as originally thought. But it looks like we can now build out _more_ files before OOM which is good.

[This PR's cache runs](https://gist.github.com/DSchau/1826ff943bc1b84694c5ea7bbb05e2a9)

[Master runs](https://gist.github.com/DSchau/918087776423ae5ac9e50aceb4e08ab5)

Basically it's looking more 1:1 than originally thought, but that's also a win if we can actually build out more things before OOM! Will keep iterating on this appraoch though, think there's some big benefits even if performance is largely the same.

**Comment by KyleAMathews** [comment]
Yeah I wouldn't expect performance to improve as reading and writing from memory is pretty fast. Our main goal is to make gatsby more scalable without hurting performance. There's other things we'll do to improve performance e.g. indexing data for queries.

**Comment by DSchau** [comment]
@KyleAMathews @pieh check this out! Now we just gotta get @pieh's working too 🙃 

This is cold build, no cache.

[Existing](https://gfycat.com/CourageousGenerousJay) | [New Approach](https://gfycat.com/PoliticalAppropriateCorydorascatfish)

This new cache is able to build out 10,000 MD files, old one fails.

**Comment by DSchau** [comment]
[Unit tests are currently failing on Node 6](https://circleci.com/gh/gatsbyjs/gatsby/4491?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

Will fix tomorrow.

File: packages/gatsby/src/utils/api-runner-node.js [hunk 5]
```diff
@@ -90,6 +90,9 @@ const runAPI = (plugin, api, args) => {
     const namespacedCreateNodeId = id => createNodeId(id, plugin.name)
 
     const tracing = initAPICallTracing(pluginSpan)
+    const cache = new Cache({
+      name: plugin.name,
+    }).init()
```
Comment (pieh) [inline packages/gatsby/src/utils/api-runner-node.js]: not sure if that matters -this will init cache for every API we run - so for markdown benchmark with 10000 pages - it will run init remark cache over 10000 times (for each `onCreateNode` run plus some single things like `createPages` etc)

Comment (DSchau) [reply packages/gatsby/src/utils/api-runner-node.js]: All I need is the plugin name. So the _earliest_ point we know that would be perfect for this PR, because then I can re-use the cache instance.

**Comment by DSchau** [comment]
@pieh and I were able to finally validate this. It seems to be slightly slower than our existing cache implementation, but not enough to not merge, because there are other benefits (e.g. cache determinism, ability to swap out stores at some point, etc.).

However, we're blocked on this until we can get a Node 6 compatible version for `node-cache-manager-fs-hash`. It uses `async/await` without a transpile step, and we can't use that in Node 6--thus the failing unit tests.

Check out the PR here rolandstarke/node-cache-manager-fs-hash#1

**Comment by KyleAMathews** [comment]
@DSchau re) your PR for node 6 support — we can always fork and vendor the code in our code base until the PR is merged/released if they don't do it promptly. Would be great to get this out ASAP.

**Comment by DSchau** [comment]
@KyleAMathews let's shoot to get this out tomorrow, so if no movement on it by then, I'll fork and create a new version?

**Comment by KyleAMathews** [comment]
👍

**Comment by DSchau** [comment]
PR was merged to `cache-manager-fs-hash`, so the unit tests should pass and we should be good here. I'll check this out more tonight/tomorrow, but I feel pretty good about it 🤞

**Comment by DSchau** [comment]
Doing some final tests on this this AM, but should be able to merge this today 🎉 

@pieh anything else I should test?

**Comment by DSchau** [comment]
Here are some more metrics.

## Sample "Real" Site (lots of queries; over 1000)

|Strategy|Cold|Warm|
|:------:|:--:|:--:|
|`cache-manager`|`86s`|`33s`|
|existing|`90s`|`37s`|

## gatsby-starter-blog

|Strategy|Cold|Warm|
|:------:|:--:|:--:|
|`cache-manager`|`7.35`|`6.10`|
|existing|`16.77`|`5.97s`|

(I'm not sure why the cold cache is so much faster 🙃)

File: packages/gatsby/src/utils/cache.js (hunk 3 above)
Comment (pieh) [reply packages/gatsby/src/utils/cache.js]: I think TTL needs to be tested - I'm not sure if we actually want this to ever expire, but from our brief testing it seemed like TTL didn't actually do anything here

Comment (DSchau) [reply packages/gatsby/src/utils/cache.js]: Yeah, so our old mechanism _never_ expired because there wasn't any TTL baked in. We could do something reasonable (like one hour?) or go a little higher... maybe 24 hours?

File: packages/gatsby/src/utils/cache.js (hunk 2 above)
Comment (pieh) [reply packages/gatsby/src/utils/cache.js]: just so it doesn't get lost - `max` is number of documents hold in memory with default config settings, but we can use additional `length` config option that could assign "weights" to documents and use some kind of heuristic to somewhat be able to limit size of in-memory cache in actual bytes to keep this under control

Comment (DSchau) [reply packages/gatsby/src/utils/cache.js]: @pieh you think it's worth experimenting with that in this PR still, or should we go ahead and merge this today sometime?

File: packages/gatsby/src/utils/cache.js (hunk 3 above)
Comment (pieh) [reply packages/gatsby/src/utils/cache.js]: this is FS store setting, right? Why should it actually expire? This will make subsequent builds (depending on where cache is used) potentially longer. For remark plugin I think it's reasonable - as build artefacts are cached as query results so caching internal data is potentially waste of disk space, but there are other plugins that use cache - i.e. wordpress plugin to know if it needs to try to redownload images or `sqip` plugin that holds results of cpu intensive tasks in cache, so even if queries need to rerun it will use cached values for `sqip` part of query results

Comment (DSchau) [reply packages/gatsby/src/utils/cache.js]: Yup. I suppose disk space is the best rationale (although I'm not sure that's a huge concern). But the key itself will be the main means of expiration, i.e. when content in the node changes, the key will inherently be new/new file/etc.

File: packages/gatsby/src/utils/cache.js [hunk 6]
```diff
@@ -4,7 +4,7 @@ const fsStore = require(`cache-manager-fs-hash`)
 const path = require(`path`)
 
 const MAX_CACHE_SIZE = 250
-const ONE_HOUR = 60 * 60
+const TTL = Number.Max_SAFE_INTEGER
```
Comment (DSchau) [inline packages/gatsby/src/utils/cache.js]: @pieh good approach?

Comment (DSchau) [reply packages/gatsby/src/utils/cache.js]: ~Whoops... needs to be all CAPS 🙃~ Fixed.

**Comment by KyleAMathews** [comment]
👍 to never clearing the cache. The cache _should_ be immutable and it's highly unlikely a site's cache will ever be larger than e.g. their node_modules directory :-) Most people have large HDs these days.

**Comment by DSchau** [comment]
@KyleAMathews @pieh this is ready whenever!

**Comment by DSchau** [comment]
Just ran it on reactjs/reactjs.org. Think this is a good reference point we can use in the future, too :)

|Strategy|Cold|Warm|
|:------:|:--:|:--:|
|`cache-manager`|`87.89`|`10.86`|
|existing|`90.84`|`11.66`|

Merging and will release 🎉

**Comment by DSchau** [comment]
```
Successfully published:
 - gatsby@2.0.18
```

**Comment by pungggi** [comment]
dont know why but upgrading to 2.0.18 gave me errors about gatsby-link missing in .cache (but it was cleared =, reverting back to 2.0.17 it worked agagin, then upgrading again to 2.0.18 worked. It happened the same whith another upgrade lately: https://github.com/gatsbyjs/gatsby/issues/8708

**Comment by DSchau** [comment]
@pungggi sorry to hear that! Did you save any of the detail in the error message?

**Comment by pungggi** [comment]
@DSchau yes here > #8708

**Comment by DSchau** [comment]
@pungggi I was confused (am confused?) because you specify 2.0.12 -> 2.0.14. If it's the same error as upgrading to that version, it seems pretty likely it's unrelated to these changes.

**Comment by pungggi** [comment]
sorry @DSchau this was not my intend obviously not right place.
